### PR TITLE
Test-FileCatalog changes to use File.OpenRead to better handle when the file is executing or open Fix #20703

### LINKED ISSUE #20703: Test-FileCatalog can't calculate a hash for an open file
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://learn.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

With Notepad++ open, open powershell and run these commands:
```
New-FileCatalog -CatalogFilePath [test.cat](http://test.cat/) -Path 'C:\Program Files\Notepad++'
Test-FileCatalog -CatalogFilePath .\[test.cat](http://test.cat/) -Path 'C:\Program Files\Notepad++'
```

At this point I get the error "Test-FileCatalog: Unable to read file C:\Program Files\Notepad++\plugins\JsonTools\JsonTools.dll to calculate its hash."

Then close Notepad++ and run 
`Test-FileCatalog -CatalogFilePath .\[test.cat](http://test.cat/) -Path 'C:\Program Files\Notepad++'`

Now it should work.

If Notepad++ is not something you can use, most any other active program seems to cause this same problem.

Here's my output.  Get-FileHash is able to generate a file hash while notepad++ is open.

```

PS C:\Users\d\Downloads\PowerShell-7.4.0-rc.1-win-x64> New-FileCatalog -CatalogFilePath [test.cat](http://test.cat/) -Path 'C:\Program Files\Notepad++'


Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
-a---          11/16/2023  1:21 PM          53728 [test.cat](http://test.cat/)

PS C:\Users\d\Downloads\PowerShell-7.4.0-rc.1-win-x64> Test-FileCatalog -CatalogFilePath .\[test.cat](http://test.cat/) -Path 'C:\Program Files\Notepad++'

Test-FileCatalog: Unable to read file C:\Program Files\Notepad++\plugins\JsonTools\JsonTools.dll to calculate its hash.

PS C:\Users\d\Downloads\PowerShell-7.4.0-rc.1-win-x64> Get-FileHash 'C:\Program iles\Notepad++\plugins\JsonTools\JsonTools.dll'


Algorithm       Hash                                                                   Path
---------       ----                                                                   ----
SHA256          E4E33D43BB06D866520C12E84F0F0760977670304DD275F5C70BE397246E1B30       C:\Program Files\Notepad++\plug…


PS C:\Users\d\Downloads\PowerShell-7.4.0-rc.1-win-x64> Test-FileCatalog -CatalogFilePath .\[test.cat](http://test.cat/) -Path 'C:\Program Files\Notepad++'

Valid


```

I've seen this on Windows 11 and Windows Server 2019.  I tested on PowerShell-7.4.0-rc.1-win-x64.  It also happens with PS 5.1.

This happens with any program. I first noticed when trying to use the catalogs to compare the differences between two servers files that the online server would fail.

### Expected behavior

```console
The Test-FileCatalog is only generating a hash with the Test-FileCatalog so it should be able to read open files.
```

### Actual behavior

```console
If file is in use it cannot create a hash.  This makes it impossible to use this to verify in use programs or files.
```

### Error details

_No response_

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.4.0-rc.1
PSEdition                      Core
GitCommitId                    7.4.0-rc.1
OS                             Microsoft Windows 10.0.19045
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### Visuals

_No response_

### PR DESCRIPTION
**Pull request by dxk3355**
Test-FileCatalog changes to use File.OpenRead Fix #20703 
Added test cases to test Test-FileCatalog with an exe that is executing 
Added test cases to test Test-FileCatalog with files with FileShare.Read and ReadWrite

# PR Summary

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: src/System.Management.Automation/security/CatalogHelper.cs [hunk 1]
```diff
@@ -423,7 +423,7 @@ internal static string CalculateFileHash(string filePath, string hashAlgorithm)
             FileStream fileStream;
             try
             {
-                fileStream = File.Open(filePath, FileMode.Open, FileAccess.Read);
+                fileStream = File.OpenRead(filePath);
```
Comment (iSazonov) [inline src/System.Management.Automation/security/CatalogHelper.cs:426]: Taking into account a nature of the issue we fix I'd prefer explicit parameters.
```suggestion
                fileStream = new FileStream(filePath, FileMode.Open, FileAccess.Read, FileShare.Read);
```

File: test/powershell/Modules/Microsoft.PowerShell.Security/FileCatalog.Tests.ps1 [hunk 2]
```diff
@@ -416,6 +416,62 @@ Describe "Test suite for NewFileCatalogAndTestFileCatalogCmdlets" -Tags "CI" {
...
+
+        AfterEach {
+            Remove-Item "$script:catalogPath" -Force -ErrorAction SilentlyContinue
+            Remove-Item "$env:TEMP\testCatalog" -Recurse -Force -ErrorAction SilentlyContinue
+        }
+
+        It "Test-FileCatalog should pass when target file has an open reader with FileMode Open FileAccess Read and FileShare Read" {
+            $script:catalogPath = "$env:TEMP\TestCatalogFileOpenValidation.cat"
+            $null = New-FileCatalog -Path "$env:TEMP\testCatalog\" -CatalogFilePath $script:catalogPath -CatalogVersion 2
+            $fileHandle = [System.IO.File]::Open("$env:TEMP\testCatalog\test.txt", [System.IO.FileMode]::Open, [System.IO.FileAccess]::Read, [System.IO.FileShare]::Read)
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Security/FileCatalog.Tests.ps1:450]: $fileStream?

File: test/powershell/Modules/Microsoft.PowerShell.Security/FileCatalog.Tests.ps1 [hunk 3]
```diff
@@ -416,6 +416,62 @@ Describe "Test suite for NewFileCatalogAndTestFileCatalogCmdlets" -Tags "CI" {
...
+        It "Test-FileCatalog should pass when target file has an open reader with FileMode Open FileAccess Read and FileShare ReadWrite" {
+            $script:catalogPath = "$env:TEMP\TestCatalogFileOpenValidation.cat"
+            $null = New-FileCatalog -Path "$env:TEMP\testCatalog\" -CatalogFilePath $script:catalogPath -CatalogVersion 2
+            $fileHandle = [System.IO.File]::Open("$env:TEMP\testCatalog\test.txt", [System.IO.FileMode]::Open, [System.IO.FileAccess]::Read, [System.IO.FileShare]::ReadWrite)
+            try{
+                $result = Test-FileCatalog -Path "$env:TEMP\testCatalog\" -CatalogFilePath $script:catalogPath
+                $result | Should -Be "Valid"
+            }
+            finally {
+                $fileHandle.Close()
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Security/FileCatalog.Tests.ps1:470]: If it is FileStream please remove.

File: test/powershell/Modules/Microsoft.PowerShell.Security/FileCatalog.Tests.ps1 (hunk 2 above)
Comment (dxk3355) [reply test/powershell/Modules/Microsoft.PowerShell.Security/FileCatalog.Tests.ps1:450]: I'm trying to create an open file to verify that the Test-Catalog can hash an open file as this was an aspect of the issue and change.  Is there a reason to not have an open stream?

File: src/System.Management.Automation/security/CatalogHelper.cs (hunk 1 above)
Comment (dxk3355) [reply src/System.Management.Automation/security/CatalogHelper.cs:426]: I was trying to be consistent with the existing code base.  Get-FileHash uses this method.  If this is the method to resolve it, I would then create an Issue that these are inconsistent.

File: test/powershell/Modules/Microsoft.PowerShell.Security/FileCatalog.Tests.ps1 (hunk 2 above)
Comment (iSazonov) [reply test/powershell/Modules/Microsoft.PowerShell.Security/FileCatalog.Tests.ps1:450]: I mean File.Open returns FileStream, not handle.

File: src/System.Management.Automation/security/CatalogHelper.cs (hunk 1 above)
Comment (iSazonov) [reply src/System.Management.Automation/security/CatalogHelper.cs:426]: I believe the source of this bug is the fact that no explicit arguments can be seen. Therefore, I ask to specify the arguments explicitly.

File: test/powershell/Modules/Microsoft.PowerShell.Security/FileCatalog.Tests.ps1 [hunk 4]
```diff
@@ -416,6 +416,62 @@ Describe "Test suite for NewFileCatalogAndTestFileCatalogCmdlets" -Tags "CI" {
...
+            $null = New-Item -ItemType Directory -Path "$env:TEMP\testCatalog" -Force -ErrorAction SilentlyContinue
+            Set-Content -PassThru "$env:TEMP\testCatalog\test.txt" -Value "Test Data" | Out-Null
+        }
+
+        AfterEach {
+            Remove-Item "$script:catalogPath" -Force -ErrorAction SilentlyContinue
+            Remove-Item "$env:TEMP\testCatalog" -Recurse -Force -ErrorAction SilentlyContinue
+        }
+
+        It "Test-FileCatalog should pass when target file has an open reader with FileMode Open FileAccess Read and FileShare Read" {
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Security/FileCatalog.Tests.ps1:447]: The two tests have identical code. Please use `-Case` parameter for FileShare Read/ReadWrite and remove second test.

Comment (dxk3355) [reply test/powershell/Modules/Microsoft.PowerShell.Security/FileCatalog.Tests.ps1]: The tests for Context "TestCatalog Skip Validation Tests" are basically having the same problem you're concerned about here as they only change a single parameter.

**Review by iSazonov — APPROVED** [review]

**Comment by iSazonov** [comment]
@dxk3355 Thanks for your contribution!

**Pull request merged**